ADP-291: Document endpoints error codes in the API documentation

hasufell opened the pull request:
# Issue Number

ADP-291

# Implementation approach

This is all best-effort. There are no static checks to ensure we didn't miss anything. There's ongoing work at servant to make this possible: https://github.com/haskell-servant/servant/issues/841
But even that would require a major refactor.

Testing all possible errors is also an option, but seems quite excessive.

The workflow for figuring out the error codes is roughly:

1. follow the entry point to the the `Handler ()` function e.g.
```hs
listTransactions
    :: forall ctx s t k n. (ctx ~ ApiLayer s t k)
    => ctx
...
    -> Handler [ApiTransaction n]
```
2. check all `liftHandler` calls, which have functions with `ExceptT` as argument, e.g.:

```hs
listTransactions
    :: forall ctx s k t.
        ( HasDBLayer s k ctx
        , HasNetworkLayer t ctx
        )
    => ctx
...
    -> ExceptT ErrListTransactions IO [TransactionInfo]
```

3. find all the LiftHandler instance, e.g.

```hs
instance LiftHandler ErrListTransactions where
    handler = \case
        ErrListTransactionsNoSuchWallet e -> handler e
        ErrListTransactionsStartTimeLaterThanEndTime e -> handler e
        ErrListTransactionsMinWithdrawalWrong ->
            apiError err400 MinWithdrawalWrong
            "The minimum withdrawal value must be at least 1 Lovelace."
        ErrListTransactionsPastHorizonException e -> handler e
```

4. follow the recursive handlers to resolve all errors and add client error types to `swagger.yaml`. The error code is the 3rd `ApiErrorCode` argument to `apiError`.
5. Some errors are implicit, e.g. failed parameter parsing etc. These are also in `ApiErrorCode`. Also see the special instance `instance LiftHandler (Request, ServerError)`
6. repeat until exhaustion

# Overview

- [x] Wallets
- [x] Addresses
- [x] Coin Selections
- [x] Transactions
- [x] Migrations
- [x] Stake Pools
- [x] Utils
- [x] Network
- [x] Proxy
- [x] Settings
- [ ] Byron-specific endpoints

# Remarks

1. I did not double check the byron endpoints. Many of their endpoints share the same types in `swagger.yaml`, so I assumed they have the same behavior wrt error codes.
2. This will generally be hard to maintain, since yaml anchors aren't enough to express the overlaps of error codes and group them sensibly. It would need something like [dhall](https://github.com/dhall-lang/dhall-lang) to do that.
3. I removed 405 errors, because the HTTP method is part of the very spec. If you follow the spec, you can't get 405.

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

Commit:
Document endpoints error codes in the API documentation

Commit:
Remove unused ErrSelectForDelegationUnableToAssignInputs

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -2261,7 +2261,6 @@ data ErrStartTimeLaterThanEndTime = ErrStartTimeLaterThanEndTime
 data ErrSelectForDelegation
     = ErrSelectForDelegationNoSuchWallet ErrNoSuchWallet
     | ErrSelectForDelegationFee ErrAdjustForFee
-    | ErrSelectForDelegationUnableToAssignInputs ErrNoSuchWallet
```
:+1: I guess that's also the case for `ErrJoinStakePoolUnableToAssignInputs` and `ErrQuitStakePoolUnableToAssignInputs`.

KtorZ commented on specifications/api/swagger.yaml:
```diff
@@ -1840,12 +1840,363 @@ x-responsesErr: &responsesErr
+  properties:
+    message:
+      type: string
+      description: A descriptive error message.
```
Instead of giving a description of the what the error message look like, wouldn't it be better to give some details about the circumstances which may lead to such error? e.g.

```
description: | 
  May occur when a given walletId does not match with any known 
  wallets (because it has been deleted, or has never existed).
```

KtorZ commented on specifications/api/swagger.yaml:
```diff
@@ -1840,12 +1840,363 @@ x-responsesErr: &responsesErr
+  properties:
+    message:
+      type: string
+      description: Explanation of a good request.
```
In a similar fashion, something that explains the root cause of the error would be better:

```
description: |
  May occur when a request is not well-formed; that is, it fails to parse
  successfully. This could be the case when some required parameters 
  are missing or, when wrong values are provided.
```

Commit:
Apply review suggestions

KtorZ committed:
adjust description of error codes to give insight about the context in which they may occur.

  Also, I've re-organized a bit their declarations so that they are listed in the same order as the branches on their corresponding sum-type in Haskell, so it's easier to map from one to another.

KtorZ committed:
remove unused error codes and error definitions from the server.

Commit:
Add automated testing for 'ApiErrorCode'

hasufell commented on lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:
```diff
@@ -917,6 +932,41 @@ spec = do
+                errStr = case res of
+                    Error s -> s
+                    _ -> ""
+            in counterexample errStr $ res == Success SchemaApiErrorCode
```
This shows a nice error message:

```
Cardano.Wallet.Api.Types
  Api Errors
    Every constructor from ApiErrorCode has a corresponding type in the schema FAILED [1]

Failures:

  test/unit/Cardano/Wallet/Api/TypesSpec.hs:937:9:
  1) Cardano.Wallet.Api.Types, Api Errors, Every constructor from ApiErrorCode has a corresponding type in the schema
       Falsifiable (after 1 test):
         Missing ApiErrorCode constructors for: ["RejectedTip","InvalidDelegationDiscovery","NotImplemented","WalletNotResponding","AddressAlreadyExists","UtxoTooSmall","WithdrawalNotWorth","PastHorizon","UnableToAssignInputOutput"]
         Each of these need a corresponding swagger type of the form: x-errConstructorName

  To rerun use: --match "/Cardano.Wallet.Api.Types/Api Errors/Every constructor from ApiErrorCode has a corresponding type in the schema/"

Randomized with seed 1387792531
```

piotr-iohk commented on specifications/api/swagger.yaml:
```diff
@@ -1887,12 +1887,505 @@ x-responsesErr: &responsesErr
+  properties:
+    message:
+      type: string
+      description: May occur when trying to forget a transaction that is not pending.
```
Just to note this will change after https://github.com/input-output-hk/cardano-wallet/pull/2262/files#diff-a2398827a22606930a9b4edc67dfc9c99196a94cdb56c56ef0ef5f92512d42e0R2467 merged. Also expired transaction will be "forgettable".

piotr-iohk commented on specifications/api/swagger.yaml:
```diff
@@ -2138,12 +2693,37 @@ x-responsesGetTransaction: &responsesGetTransaction
+        schema:
+          oneOf:
+            - <<: *errBadRequest
+            - <<: *errTransactionIsTooBig
```
I think this should be only in `403`?
![Screenshot from 2020-10-30 11-10-06](https://user-images.githubusercontent.com/42900201/97692793-7f6ef600-1aa0-11eb-82eb-a5765196ce7b.png)

hasufell replied:
```
        ErrSelectForPaymentTxTooLarge maxSize maxN  ->
            apiError err400 TransactionIsTooBig $ mconcat
                [ "I am afraid that the transaction you're trying to submit is "
                , "too large! The network allows transactions only as large as "
                , pretty maxSize, "s! As it stands, the current transaction only "
                , "allows me to select up to ", showT maxN, " inputs. Note "
                , "that I am selecting inputs randomly, so retrying *may work* "
                , "provided I end up choosing bigger inputs sufficient to cover "
                , "the transaction cost. Alternatively, try sending to less "
                , "recipients or with smaller metadata."
                ]
```
it is 400

piotr-iohk commented on specifications/api/swagger.yaml:
```diff
@@ -2138,12 +2693,37 @@ x-responsesGetTransaction: &responsesGetTransaction
+    content:
+      application/json:
+        schema:
+          <<: *errNoSuchWallet
```
Same `410` and errNoSuchWallet.

hasufell replied:
It is 410

```
        ErrSignPaymentNoSuchWallet e -> (handler e)
            { errHTTPCode = 410
            , errReasonPhrase = errReasonPhrase err410
            }
```

piotr-iohk commented on specifications/api/swagger.yaml:
```diff
@@ -1908,19 +2401,12 @@ x-responsesErr404: &responsesErr404
       application/json:
         schema: *responsesErr
 
-x-responsesErr405: &responsesErr405
```
Do we think it is irrelevant to highlight in API doc? Basically every endpoint may return this...

hasufell replied:
I don't think so. The spec includes which method to use. If you follow the spec, 405 will never be returned.

piotr-iohk replied:
Then `406` and `415` should be removed also I guess, no?

hasufell replied:
415 maybe, but the swagger spec doesn't document enough for 406 to not be a concern afais.

piotr-iohk commented on specifications/api/swagger.yaml:
```diff
@@ -2255,22 +2869,67 @@ x-responsesListStakePools: &responsesListStakePools
+    content:
+      application/json:
+        schema:
+          <<: *errNoSuchWallet
```
Same `410` and errNoSuchWallet.

hasufell replied:
```
        ErrSignPaymentNoSuchWallet e -> (handler e)
            { errHTTPCode = 410
            , errReasonPhrase = errReasonPhrase err410
            }
```

piotr-iohk commented on specifications/api/swagger.yaml:
```diff
@@ -2065,11 +2582,24 @@ x-responsesGetWalletMigrationInfo: &responsesGetWalletMigrationInfo
+    content:
+      application/json:
+        schema:
+          <<: *errNoSuchWallet
```
This doesn't feel right to me. I think we should use `404` on 'no such wallet'. Especially looking at some endpoints where:

![Screenshot from 2020-10-30 11-25-22](https://user-images.githubusercontent.com/42900201/97694299-a1697800-1aa2-11eb-8c4c-d16402df5a71.png)

hasufell replied:
This PR isn't really about refactoring or fixing our error codes though. I just added documentation to the existing error handlers. I changed nothing.

piotr-iohk replied:
ACK. It revealed those small inconsistencies though (I spotted 410 and 403 only). 
Could be fixed in this go if it's not a lot of work, but don't have to of course.

piotr-iohk commented on specifications/api/swagger.yaml:
```diff
@@ -2255,22 +2869,67 @@ x-responsesListStakePools: &responsesListStakePools
+    content:
+      application/json:
+        schema:
+          <<: *errNoSuchWallet
```
Same `410` and errNoSuchWallet.

hasufell replied:
```
        ErrSignPaymentNoSuchWallet e -> (handler e)
            { errHTTPCode = 410
            , errReasonPhrase = errReasonPhrase err410
            }
```

piotr-iohk reviewed (commented):
Good work on documenting error messages. I've left few comments.
